metadata-io/src/main/java/com/linkedin/metadata/entity/DeleteEntityService.java

```diff
@@ -0,0 +1,254 @@
+
+  /**
+   * Public endpoint that deletes references to a given urn across DataHub's metadata graph. This is the entrypoint for
+   * addressing dangling pointers whenever a user deletes some entity.
+   *
+   * @param urn     The urn for which to delete references in DataHub's metadata graph.
+   * @param dryRun  Specifies if the delete logic should be executed to conclusion or if the caller simply wants a
+   *                preview of the response.
+   * @return  A {@link DeleteReferencesResponse} instance detailing the response of deleting references to the provided
+   * urn.
+   */
+  public DeleteReferencesResponse deleteReferencesTo(final Urn urn, final boolean dryRun) {
+    final DeleteReferencesResponse result = new DeleteReferencesResponse();
+    RelatedEntitiesResult relatedEntities =
+        _graphService.findRelatedEntities(null, newFilter("urn", urn.toString()), null,
+            EMPTY_FILTER,
+            ImmutableList.of(),
+            newRelationshipFilter(EMPTY_FILTER, RelationshipDirection.INCOMING), 0, 10000);
+
+    final List<RelatedAspect> relatedAspects = relatedEntities.getEntities().stream().map(relatedEntity -> {
+          final Urn relatedUrn = UrnUtils.getUrn(relatedEntity.getUrn());
+          final String relationType = relatedEntity.getRelationshipType();
+          final String relatedEntityName = relatedUrn.getEntityType();
+          final EntitySpec relatedEntitySpec = _entityService.getEntityRegistry().getEntitySpec(relatedEntityName);
+
+          return getAspectReferringTo(urn, relatedUrn, relationType, relatedEntitySpec).map(pair -> {
+            final RelatedAspect relatedAspect = new RelatedAspect();
+            relatedAspect.setEntity(relatedUrn);
+            relatedAspect.setRelationship(relationType);
+            relatedAspect.setAspect(pair.getSecond().getName());
+            return relatedAspect;
+          });
+        }).filter(java.util.Optional::isPresent)
+        .map(java.util.Optional::get)
+        .limit(10).collect(Collectors.toList());
+
+    result.setRelatedAspects(new RelatedAspectArray(relatedAspects));
+    result.setTotal(relatedEntities.getTotal());
+
+    if (dryRun) {
+      return result;
+    }
+
+    int processedEntities = 0;
+
+    // Delete first 10k
+    relatedEntities.getEntities().forEach(entity -> deleteReference(urn, entity));
+
+    processedEntities += relatedEntities.getEntities().size();
+
+    // Delete until less than 10k are left
+    while (relatedEntities.getCount() > 10000) {
+      log.info("Processing batch {} of {} aspects", processedEntities, relatedEntities.getTotal());
+      sleep(ELASTIC_BATCH_DELETE_SLEEP_SEC);
+      relatedEntities = _graphService.findRelatedEntities(null, newFilter("urn", urn.toString()),
+          null, EMPTY_FILTER, ImmutableList.of(),
+          newRelationshipFilter(EMPTY_FILTER, RelationshipDirection.INCOMING), 0, 10000);
+      relatedEntities.getEntities().forEach(entity -> deleteReference(urn, entity));
```

Comment:
What happens if deleteReference fails? Do we catch this, retry, etc? or does it just break the entire cleanup process?

Reply:
`What happens if deleteReference fails?`
We log the error and continue the delete process for the other related entities.

If there is a runtime exception then the cleanup process is broken.

Reply:
Maybe we need to explicitly log to a dead-event queue? Basically I way to flag the fact that we missed deleting something as LOUDLY and clearly as possible. This is highly risky - we don't want to end up in a half consistent state, end up making assumptions elsewhere in the system, and kind of do one of these: 🤷